File: go/internal/coagulator1d/sequential/coagulator.go
package sequential

import (
	"github.com/maxkuzn/advection-and-coagulation/algorithm/coagulation"
	"github.com/maxkuzn/advection-and-coagulation/internal/field1d"
)

type coagulator struct {
	base coagulation.Coagulator
}

func New(base coagulation.Coagulator) *coagulator {
	return &coagulator{base: base}
}

func (c *coagulator) Start() error {
	return nil
}

func (c *coagulator) Stop() error {
	return nil
}

func (c *coagulator) Process(field, buff field1d.Field) (field1d.Field, field1d.Field) {
	for i := 0; i < field.Len(); i++ {
		c.base.Process(field.Cell(i), buff.Cell(i), field.Volumes())
	}

	return field, buff
}

File: cpp/library/coagulator1d/sequential/coagulator.cpp
#include "coagulator.h"

#include <memory>

namespace coagulation {

std::pair<Field1D*, Field1D*> SequentialCoagulator1D::Process(Field1D* field, Field1D* buff) {
	Field1D& f = *field;
	Field1D& b = *buff;

	for (size_t i = 0; i < f.Size(); i++) {
		base_coagulator_->Process(&f[i], &b[i], f.Volumes());
	}

	return std::make_pair(field, buff);
}

};  // namespace coagulation
